### ISSUE TITLE
[astro] missing events

### ISSUE THREAD

**Issue by udo1toni** [openhab/openhab-addons#16565]
## Expected Behavior
As far as I understand, every rangeEvent channel which triggers START, should also trigger END

## Current Behavior
This seems to be true for most rangeEvents, but not for `eveningNight` and `night`. Both channels do only trigger `START`, but not `END`.

## Steps to Reproduce (for Bugs)
1. setup a thing with "common" latitude/longitude (so it is to be expected to get eveningNight and night triggers)
2. wait a day or so
3. get event logs for the channels

## Possible solution
The moment in time for each trigger is calculated at midnight for the upcoming day. 
But the former event is still "active". So maybe calculation for night/eveningNight event has to be done in a different way (e.g. don't change END time until the event got to END)

## Your Environment

* openHAB4.1.1 release build
* openJDK17 headless
* Debian12-amd64 as LXContainer (Proxmox 8.1.5)

See https://openhabforum.de/viewtopic.php?p=60851#p60851 for reference

**Comment by Oekel** [comment]
Reproduceable

**Comment by lsiepel** [comment]
`eveningNight`  always ends at 00:00 `morningNight` starts at 00:00.  `night` covers both together.
I just checked and the ranges are calculated correctly.  When scheduling the events, only the same day events are scheduled. 

This is [filtered here](https://github.com/openhab/openhab-addons/blob/104f32892e87d3c5236dad7326d10cad4b15f6df/bundles/org.openhab.binding.astro/src/main/java/org/openhab/binding/astro/internal/job/Job.java#L55)

So i would only expect the `night` event not to end. As a workaround you can use the `morningNight` End, it is the same moment in time.

You also mention the `eveningNight` not raising end. This is probably be because it is record at 00:00 midnight and is also regarded as a next day event. Some sort of race condition.

I wonder why there is a sameday filter. As removing that would fix this issue, but could also cause some weird behaviour as events are stacked?!

**Comment by lsiepel** [comment]
@nadahar would this possibly be fixed with the astro PR you created the past week?

**Comment by Nadahar** [comment]
> would this possibly be fixed with the astro PR you created the past week?

One should never say never, but I kind of doubt it. My work was focused on handling timezones and locales properly, and on the tests not failing randomly. It sounds like this is a problem more at "the logic level", which I haven't really touched.

**Comment by Markkuuss** [comment]
I can confirm this behavior and have some debug logs that may help identify the root cause.

For me, not only the `END` event is missing. Both `START` and `END` events for `night` are not scheduled at all.

Debug log:

```text
2026-06-10 00:00:20.325 [DEBUG] [org.openhab.binding.astro.internal.job.Job] - Not scheduling Event job astro:sun:HomeEvents/night#event/START since it's at another date (Thu Jun 11 00:25:16 CEST 2026)

2026-06-10 00:00:20.326 [DEBUG] [org.openhab.binding.astro.internal.job.Job] - Not scheduling Event job astro:sun:HomeEvents/night#event/END since it's at another date (Thu Jun 11 01:59:47 CEST 2026)
```

This seems to support the theory that range events spanning midnight are not handled correctly.

At midnight on June 10th, the Astro binding calculates the `night` range for the following calendar day:

* START: Jun 11, 00:25
* END: Jun 11, 01:59

Both events are then skipped because they are considered to be "at another date".

From the log it looks like the scheduler only creates jobs for events occurring on the current calculation date and discards events that fall on the next calendar day, even though they belong to the active `night` range.

Could this be the reason why `night` and possibly other range events crossing midnight never generate their expected events?

**Comment by Nadahar** [comment]
> I can confirm this behavior and have some debug logs that may help identify the root cause.
> 
> For me, not only the `END` event is missing. Both `START` and `END` events for `night` are not scheduled at all.
> 
> Debug log:
> 
> ```
> 2026-06-10 00:00:20.325 [DEBUG] [org.openhab.binding.astro.internal.job.Job] - Not scheduling Event job astro:sun:HomeEvents/night#event/START since it's at another date (Thu Jun 11 00:25:16 CEST 2026)
> 
> 2026-06-10 00:00:20.326 [DEBUG] [org.openhab.binding.astro.internal.job.Job] - Not scheduling Event job astro:sun:HomeEvents/night#event/END since it's at another date (Thu Jun 11 01:59:47 CEST 2026)
> ```
> 
> This seems to support the theory that range events spanning midnight are not handled correctly.

The original issues have been fixed a long time ago. It is correct (ish) to skip those events (I no longer remember the details, but I thought we kept events for the next 26 hours, which means that they probably shouldn't have been skipped). Regardless, that's not the main problem here. The main problem is why the "night" that is found is the "next one", not the "current one". Also, for the night to be this short, you must live up here in Norway or Sweden?

My guess is that there's an issue with the timezone setup, that it doesn't match the long/lat that is configured. The long/lat is used to calculate the "movement" of the sun, while the timezone determines how this fits within "the next 26 hours" window. Also, what version of OH are you running?

**Comment by Markkuuss** [comment]
> Also, for the night to be this short, you must live up here in Norway or Sweden?
> 
> My guess is that there's an issue with the timezone setup, that it doesn't match the long/lat that is configured. The long/lat is used to calculate the "movement" of the sun, while the timezone determines how this fits within "the next 26 hours" window. Also, what version of OH are you running?

No, I'm located in Germany. The timezone and coordinates appear to be configured correctly.

<img width="739" height="346" alt="Image" src="https://github.com/user-attachments/assets/f1df370a-ed01-45bf-ba7d-37743d12e610" />

I'm running the current release: openHAB 5.1.4

**Comment by Nadahar** [comment]
I don't remember how much was backported or not, but I don't think that the scheduling fixes I made has been backported to 5.1. When I check my current log for 5.2.0, I find entries like this:

```
02:09:06.784 [DEBUG] (OH-safeCall-1       ) [enhab.binding.astro.internal.job.Job] - Not scheduling Event job astro:sun:local/eclipse#event/PARTIAL because it's outside the schedulable time window (2029-01-14T18:13:34.973947525+01:00[Europe/Oslo])
```

Please note that the text is different, instead of "since it's at another date" it is "because it's outside the schedulable time window". I think this text was changed when I refactored the scheduling, which is what fixed "night", "midnight" and a number of other issues. So, you probably must still wait for 5.2.0 to get a version where this is working.

That said, it seems like there is some issue with the version I'm running as well, because I currently have no "night" event (or both endpoints are `NULL`). But, that might be "correct" because of the time of the year and my location. There is no real "night" here these days, nautic dusk starts at 23:54 but never ends, nautic dawn never starts, but ends 02:45. There are no astro dusk or down, it never gets dark enough/the sun never is low enough for that. Since "night" is between astro dusk and astro dawn, it "disappears" as well. It's still a bit "strange" in that it's relatively dark outside at the moment, and I would "call it" night - but I guess that astronomically, it isn't.

All in all, you'll either have to wait for 5.2.0 (shouldn't be too far off now), or see of you can get a snapshot version of the Astro binding working with 5.1.4.

**Comment by Nadahar** [comment]
@lsiepel I *think* this can be closed when 5.2.0 is released. I'm not quite sure how you do that, you can't really close issues that are solved in the latest snapshot months before it's released, or?

**Comment by lsiepel** [comment]
We close them when the PR is merged. What PR are you referring to?

**Comment by Nadahar** [comment]
> What PR are you referring to?

Don't remember, don't even remember if it was only one PR, but I pretty much refactored the whole event scheduling. After that, some "Instance refactoring" was done, but I don't know how much of it, if any, that broke. This particular problem should be solved anyway, because I changed it from scheduling "within the same day" to scheduling for the next 26 hours. That means that some events could be double scheduled, so I made some "deconflicting" logic to handle that. This fundamentally changed the problems surrounding events at or soon after when the event recalculation itself is scheduled (soon after midnight). I also made some handling of the fact that the night might extend into the past (it's already night) or beyond the schedule window.

As a result, I'm pretty sure that the issues discussed here are no more, but it's hard to properly verify that until the refactored code is in the hand of the users.

**Comment by lsiepel** [comment]
ok, i'll close this for now. When one of the above users can reproduce this on the latest snapshots we can reopen, but let's close this for now.
